### ISSUE TITLE
Question about 'wmdrun' and 'workers' in the infretis config

### ISSUE THREAD

**Issue by hhlim12** [infretis/infretis#123]
Hi, thank you for developing very nice package.

I'd like to run infretis with LAMMPS.
Would you mind to answer the following question?

- Could you tell me the meaning of `wmdrun` in the config file? How will it be used in the MD execution?
- How to find the optimum number of workers? Is it correct the bigger the number gives better efficiency?

Thanks in advance.

**Comment by lukasbaldauf** [comment]
Hi,
thank you for the kind words!

> Could you tell me the meaning of wmdrun in the config file? How will it be used in the MD execution?

The `wmdrun` setting can set specific worker options for running the MD. This is mainly used with GROMACS, when one wants to specify a slightly different execution command for each worker.

For example,
```toml
workers = 2
wmdrun = [
	"gmx mdrun -option1",
	"gmx mdrun -option2",
]
```
These commands are then set in the `set_mdrun()` function within `infretis/infrretis/classes/engines/gromacs.py`, such that each of the two worker invokes GROMACS with one of the specified mdrun commands, thereof the name "wmdrun", for "worker mdrun".

For LAMMPS however, there are usually no worker-specific options to set. If you look at the `set_mdrun()` function for the LAMMPS engine, it doesn't use the `wmdrun` values at all. But i believe it gives an error if they are not set, so one must supply a list with a length equal to the number of workers, where each element contains some string (see `examples/lammps/H2/infretis.toml`).

> How to find the optimum number of workers? Is it correct the bigger the number gives better efficiency?

@sroet and @dz24 Correct me if I'm wrong here:

That number can be a bit difficult to set. It depends on the process you are studying, the hardware at hand, and also how fast you want your results.

Usually, one beforehand specifies the number of workers by considering the available hardware and the performance. Say you have 32 CPUs available, and the performance of one MD simulation doubles when you double the number of CPUs up to 4 CPUs, after which the scaling becomes worse. Then, if you check the performance of running 8 concurrent MD simulations with 4 CPUs each, and you get a total performance that is 8 times the performance of a single worker, I would pick 8 workers. 

Then, we usually pick the number of interfaces that is twice (plus 1 interface) the number of workers, so 16+1 = 17 interfaces. 

You could then say that these are your final worker settings. However, it also depends on the process you are studying.

If it is very rare, you might want to have more interfaces such that you at a local crossing probability of around 0.3, meaning that a 30% of paths in a given ensemble $i$ cross the next interface $i+1$. If you then have access to another node, you could then consider using a second node, giving 16 workers with 33 interfaces. Then, however, one can run into storage problems if your trajectories are really large :elephant:

If the process is not so rare, and your resulting local crossing probability is much greater than 0.3, e.g. 0.7 or so, it might not be necessary to have that many workers (and interfaces), depending on how fast you want your answers.

However, the local crossing probability is not a number one knows beforehand, but one can make educated guesses based on short infretis or MD runs, or free-energy profiles.


I hope this helps, let me know if you have any other questions :smile: :infinity:

**Comment by hhlim12** [comment]
Thanks so much for the detail response! It helps a lot.

Now I understand about the wmdrun. I use LAMMPS so then I will not need it.

Regarding the number of workers, my situation is that my MD simulations are rather slow and the performance declines at 32 cpus. I can afford more cpus, so I'm thinking to make as many workers as possible. Then this results to very very dense interfaces even though the process is not so rare. I wonder whether it will give redundant samples due to the very small differences in the order-parameter. Also, I have tried some examples before and found that small number of interfaces converge faster..Do you have any comment on that?

**Comment by lukasbaldauf** [comment]
I feel like I'm not the most qualified to answer this, maybe @sroet or @dz24 could elaborate. A detailed analysis was performed in [J. Phys. Chem. A 2022, 126, 8878−888](https://doi.org/10.1021/acs.jpca.2c06004).

But two practical things I could think of:
*  The more workers you add, the more initial paths you need. If many of these initial paths are similar, or maybe even identical, then the convergence could be slowed down because the paths have to decorrelate.

* Another thing is to make sure that the interfaces are not placed too far towards state B. In that case, you would end up with a lot of B to B paths (look for status _BWI_ in the sim.log file).  Here, you could experiment with the _interface_cap_ keyword in the .toml, and set it such that you don't get a lot of these B->B paths. Then, the next-to-largest interface should be placed a bit lower than this cap, to avoid situations where we have few phasepoints between the interface sand the cap. See the image below, where interfaces are in black, and the interface cap is in blue.

![image](https://github.com/user-attachments/assets/1083f010-999e-47a9-9929-446a783fc431)
* After the WHAM analysis, make sure you are using the column that used WHAM'ing in the output files, as single-point matching would be much less effective here when there is large overlap between interfaces.

**Comment by dz24** [comment]
I think your advice is good @lukasbaldauf.

> I wonder whether it will give redundant samples due to the very small differences in the order-parameter

If you can afford more cpus, then, generally, initiating a larger number of workers would help converge the results faster. It would not necessarily generate redundant samples, rather you get extra samples faster, but the paths may have to become de-correlated from the initial load trjectories first (through series of shooting moves).

**Comment by sroet** [comment]
I also agree with the advice of @lukasbaldauf !

>I feel like I'm not the most qualified to answer this, maybe @sroet or @dz24 could elaborate. A detailed analysis was performed in [J. Phys. Chem. A 2022, 126, 8878−888](https://doi.org/10.1021/acs.jpca.2c06004).

Just to point you to the correct part of that reference; the theoretical efficiency estimations are in part 6 of the SI.

The last time I had a discussion on this (more than a year ago) the optimal number of interfaces is hard to determine beforehand, but the best bet was to always use double the number of workers (+1) as interfaces. However, as @lukasbaldauf already mentioned, you should reach at least 0.2 crossing probability on every interface otherwise you should use more.  
I also agree that any interface that has a crossing probabilty > 0.7-0.8 should either be removed or moved between two interfaces that have a crossing probability < 0.5 and you shouldn't place them too close to B

To comment on the maximum number of workers (assuming you don't mind wasting resources), there is a point where the algorithm becomes slower (less walltime efficient) for many workers (see for example figure 1F of the reference)
The is due to the fact that computation of the infinite swapping which happens after a worker is done with one move and before starting with the next one is **almost free**, but not completely. My initial implementation (@dz24 or @lukasbaldauf can comment on the current one), aimed to do it under 1 second. This has a couple implications:

1) The used algorithm could reliably do it under 1 second for up to **3500** free interfaces, so ~3500 workers/7000 interfaces if you use our double the number of workers as number of interfaces, **if you don't use high-acceptance moves** (see section 4 of the SI of the referenced paper). You are free to use more than 3500 workers, but just make sure **#ensembles-#workers <= 3500**

If you do use high-acceptance then point 1 becomes dependent on the crossing-probability, as the algorithm then only can do blocks of 12x12 in the W matrix under a second. This means there is a probability of not being able to do it under a second of on average $$((N-(W-1))-12)((P^{\frac{N}{N-(W-1)}*{12}})$$ where $N$ is the total number of interfaces $W$ is the number of workers and $P$ is the crossing probability (assuming all interfaces have the same crossing probability). So again the solution would then be to increase the ratio of workers versus the number of ensembles. However in my initial implementation (don't know if that is still present), my code would switch to doing 100.000 swaps instead of doing infinite swaps on that block reducing sampling efficiency for a more guaranteed wall-time behavior, and I assume something like that is present still.

2) The 1 second estimate also means that if  **(average MD walltime per MC move)/(# workers)** approaches **100 s** you start spending more than 1% of your simulation time on the swapping instead of generating more MD, which could actively make your simulation slower if it approaches the 1-10 second range (the graph of figure 1F was for a test system where the average MD walltime per MC move was already ~1s for a single worker, resulting in an optimal number of workers of **2**) 

The above points are just the algorithm overhead, not considering the communication overhead between the MD worker and the main one, or storage speeds which might put the maximum number of workers (way) earlier than your average MD/#workers reaching the 10-100s range.

**Comment by hhlim12** [comment]
Thanks so much for very detailed answer!. 
I'm still digesting the detailed analysis in the paper. 
Based on the discussion, I have done the following.
- Increase the number of workers for my simulation with 2n+1 number for interfaces.
- Ensure the initial load trajectory are decorrelated.
- Ensure the local probability for each interface is not too small.

I also have performed the WHAM analysis using the `inftools` package.
However, the current output of WHAM analysis is quite difficult to understand. 
Do you have notes or brief explanation for the files?
For instance, I got these files.

<img width="1569" alt="image" src="https://github.com/user-attachments/assets/0f5f4c02-f9ac-41bb-9880-24bcc1f93b88" />

I think the local probability should be in the ploc_WHAM. But the meaning of each column is not clear for me.
The header of the file show something like:
```
#Lam    [0+]    [1+]    [2+]    [3+]    [4+]    [5+]    [6+]    [7+]    [8+]    [9+]    [10+]   [11+]   [12+]   [13+]   [14+]
-6.0    1.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0
-5.9    0.5942404875495252      0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0
-5.800000000000001      0.3016448192710098      0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0
```
Which one should I plot? and how to get the plot like this? Do you have the formula written in the paper?

<img width="679" alt="image" src="https://github.com/user-attachments/assets/fb8c710c-89a1-4d07-8cfa-0c1786ec0dc7" />

Thanks so much for your kind attention..

**Comment by lukasbaldauf** [comment]
Hi, the plot was just a simple double well for illustration purposes. I’m in the process of writing something on this. For now, it is available here:

https://github.com/infretis/infentory/blob/infretis_info/chignolin/infretis_output.md

I’m updating it these days, so there will come some more detailed information.

**Comment by hhlim12** [comment]
Thanks so much for your help.
I can plot the local probabilities and total crossing probability now.

Here is my current results.
It seems my crossing probabilities are too small for some interfaces.

<img width="578" alt="image" src="https://github.com/user-attachments/assets/bf3aec93-6827-43b2-8e5d-ebdc397853ad" />

<img width="568" alt="image" src="https://github.com/user-attachments/assets/579e51ea-14f4-421a-893a-837c830d044c" />

Thanks again for answering my questions!.

**Comment by lukasbaldauf** [comment]
> Here is my current results. It seems my crossing probabilities are too small for some interfaces.
> 
> <img alt="image" width="568" src="https://private-user-images.githubusercontent.com/77390746/395746471-579e51ea-14f4-421a-893a-837c830d044c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQxNjA4NjIsIm5iZiI6MTczNDE2MDU2MiwicGF0aCI6Ii83NzM5MDc0Ni8zOTU3NDY0NzEtNTc5ZTUxZWEtMTRmNC00MjFhLTg5M2EtODM3YzgzMGQwNDRjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMTQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjE0VDA3MTYwMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWY5NmEyY2I0YWYwMjdjMzFiMGU5ZjRiN2RiMGY2MzM0ODc5NzhlMGJiYjA5MmNhMjAzYTRhNmEyNGFhZDY2NDQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.MyuQvcge0kAEz60YFthhzo1Ud-WvfTYtJnmqmP9dkgY">

Just for clarity, this plot looks like the *scaled* local crossing probabilities. To check the interface placement one should use the unscales probabilities from the ‘ploc_unscaled.txt’ file, where  the probability for ensemble [i+] begins with a value of 1.0 at interface $i$. So each of the colored lines in that last plot should begin at 1.0.

**Comment by hhlim12** [comment]
Thanks so much! I replot using the `ploc_unscaled.txt` and now it shows higher probability which is more reasonable.
<img width="1767" alt="image" src="https://github.com/user-attachments/assets/cd48513f-7480-47da-acc6-036429f9ac89" />

>Also, the latter interfaces are redundant and should be removed or placed elsewhere in an eventual next run, as all paths generated in last region where the total crossing probability is flat would be rejected because they are B->B paths.
>One can use inftools “inft estimate_interfaces2” to estiamte interface locations based on the total crossing probability (Pcross.txt) if you plan to run longer.

I will definitely try this. Thanks again!

**Comment by hhlim12** [comment]
Hi, Thanks a lot for your help. I have tried the `inft estimate_interfaces2` and it gives me the interface positions.

`inft estimate_interfaces2 -i Pcross.txt -ploc 0.2`

The output is as follows.
`[INFO] Added 11 interfaces
interfaces = [ -6.000000, -5.700000, -2.400000, -2.200000, -2.100000, -2.000000, -1.900000, -1.800000, -1.700000, -1.600000, 3.800000]`

Just to make sure, does the output gives me the "additional interface to my original ones" or the "all actual interfaces I should use" ? Because "Added 11 interfaces" kind of give me the former one is correct.

Also, should I throw away my old calculations when I change the interfaces?

Sorry for asking so many questions. Really appreciate your help!

**Comment by lukasbaldauf** [comment]
No problem, it's nice to have these discussions here because there isn't any information about this out yet. So don't worry about asking a lot of questions.

> Just to make sure, does the output gives me the "additional interface to my original ones" or the "all actual interfaces I should use" ? 

The output would be "all the actually interfaces you should use", so the print statement could be a bit more clear on that. It is more clear when you also give the -plot option.

For example,
```bash
inft estimate_interfaces2 -i Pcross.txt -ploc 0.2 -i0 -6.0 -iN 4.0 -plot
```
gives 

![nocap](https://github.com/user-attachments/assets/0dda3aeb-4b7e-4a32-bed6-486272002e6b)

Another way that might also be useful when using a larger value for ploc with many workers: 
* set -iN to the value of your interface cap
* and then remove the last value from the output list and place it as the _interface_cap_ parameter in the .toml

For example, assuming you set an _interface_cap_ at -1.4:
```bash
inft estimate_interfaces2 -i pcross.txt -ploc 0.4 -i0 -6.0 -iN -1.4 -plot
```
Outputs
```text
[INFO] Added 21 interfaces
interfaces = [ -6.000000, -5.87, -5.73, -5.16, -2.42,
-2.22, -2.15, -2.09, -2.03, -1.99, -1.95, -1.91, -1.89,
-1.87, -1.85, -1.80, -1.74, -1.69, -1.65, -1.61, -1.40000]
```

Then in the .toml
```toml
interface_cap = -1.4
interfaces = [ -6.000000, -5.87, -5.73, -5.16, -2.42,
-2.22, -2.15, -2.09, -2.03, -1.99, -1.95, -1.91, -1.89,
-1.87, -1.85, -1.80, -1.74, -1.69, -1.65, -1.61,
4.00]  # note -1.4 replaced with 4.0
```
> Also, should I throw away my old calculations when I change the interfaces?

Your curve looks rather smooth, so it might not be worth rerunning it, as the results you now have aren't wrong. But if you want to reduce the overall error and you want very accurate results, it may be worthwhile.

The old results can then not be used directly.

We are working with an automated procedure called `infinit` that does this interface placement behind the scenes, so one doesn't have to worry about it at all. But it's still work-in-progress.

**Comment by hhlim12** [comment]
Thank you very much for replying! I decide to rerun the simulation to reduce the overall error. 
I'd like to confirm, the unit of rel-error in "errPtot.txt" and "errploc.txt" should be in % right?

**Comment by dz24** [comment]
(Copy paste from someone asking a similar question via email)

The errors are relative errors, so an error of X means that the estimated error is X times the estimated value.

It is computed from block-averaging and at the top of the file it writes the average of the block-errors (estimates of error at different block-lengths).

**Comment by hhlim12** [comment]
I understand. Thank you very much for answering. I guess now I can close this issue..
